data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java

```diff
@@ -0,0 +1,128 @@
+ * In order to do so, this component:
+ * <ul>
+ *   <li>Starts an {@link HttpServer} listening for incoming events</li>
+ *   <li>Starts a {@link ResourcesReconciler}, listen on the event bus for reconciliation events and keeps track of the {@link dev.knative.eventing.kafka.broker.contract.DataPlaneContract.Ingress} objects and their {@code path => (topic, producer)} mapping</li>
+ *   <li>Implements a request handler that invokes a series of {@code preHandlers} and then a final {@link IngressRequestHandler} to publish the record to Kafka</li>
+ * </ul>
+ */
+public class ReceiverVerticle extends AbstractVerticle {
+
+  private static final Logger logger = LoggerFactory.getLogger(ReceiverVerticle.class);
+
+  private final HttpServerOptions httpServerOptions;
+  private final Function<Vertx, IngressProducerReconcilableStore> ingressProducerStoreFactory;
+  private final Iterable<Handler<HttpServerRequest>> preHandlers;
+  private final IngressRequestHandler ingressRequestHandler;
+
+  private HttpServer server;
+  private MessageConsumer<Object> messageConsumer;
+  private IngressProducerReconcilableStore ingressProducerStore;
+
+  public ReceiverVerticle(HttpServerOptions httpServerOptions,
+                          Function<Vertx, IngressProducerReconcilableStore> ingressProducerStoreFactory,
+                          Iterable<Handler<HttpServerRequest>> preHandlers,
+                          IngressRequestHandler ingressRequestHandler) {
+    Objects.requireNonNull(ingressProducerStoreFactory);
+    Objects.requireNonNull(ingressRequestHandler);
+
+    this.httpServerOptions = httpServerOptions != null ? httpServerOptions : new HttpServerOptions();
+    this.ingressProducerStoreFactory = ingressProducerStoreFactory;
+    this.preHandlers = preHandlers != null ? preHandlers : Collections.emptyList();
+    this.ingressRequestHandler = ingressRequestHandler;
+  }
+
+
+  @Override
+  public void start(final Promise<Void> startPromise) {
+    this.ingressProducerStore = this.ingressProducerStoreFactory.apply(vertx);
+    this.messageConsumer = ResourcesReconciler
+      .builder()
+      .watchIngress(this.ingressProducerStore)
+      .buildAndListen(vertx);
+
+    this.server = vertx.createHttpServer(httpServerOptions);
+
+    this.server.requestHandler(this::handle)
+      .exceptionHandler(startPromise::tryFail)
+      .listen(httpServerOptions.getPort(), httpServerOptions.getHost())
+      .<Void>mapEmpty()
+      .onComplete(startPromise);
+  }
+
+  @Override
+  public void stop(Promise<Void> stopPromise) {
+    CompositeFuture.all(
+      server.close().mapEmpty(),
+      messageConsumer.unregister()
+    )
+      .<Void>mapEmpty()
+      .onComplete(stopPromise);
+  }
+
+  private void handle(HttpServerRequest request) {
+    // Run preHandlers
+    for (Handler<HttpServerRequest> preHandler : this.preHandlers) {
+      preHandler.handle(request);
+      // preHandler might end the request prematurely
+      if (request.isEnded()) {
```

Comment:
What if a preHandler handles a request asynchronously, would this return false?
It seems we can a build a more robust design here :thinking:

Reply:
Do we have an actual use case for that? I feel like I'm recreating `vertx-web`, and I don't wanna go down that route :smile: How about we introduce that change only if a use case comes out?

Reply:
A class shouldn't change because another one changes. 
Why don't we organize these handlers as a chain https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern?

Reply:
> A class shouldn't change because another one changes.

Of course, what I meant is that because we have no use case for async pre handlers, nor I can think of any future use cases in mind for that, there is no need for `ReceiverVerticle` to support async handlers, because it makes the code more complicated for no reason. 

What I'm implementing here is already a chain of responsability, where handlers are invoked one after the other where each one does its job, It's just not meant to include async handlers. In order to implement it with async handlers, we should have `Function<HttpServerRequest, Future<Void>>`.

The reason why the `IngressRequestHandler` is "special" is because I would like to have the implementation be unaware of the store, the reconciliation and all the complexity brought in with it. I tried to find a nice way to fix that, but then the injection of the `vertx` object would become too much complicated, so I'm out of solutions atm :disappointed: 

The solution in this PR is not be best solution of course, but I guess it's a beginning...

Reply:
I see, let's keep this for now